### ISSUE TITLE
Possibility to choose Graphics backend at runtime

### ISSUE THREAD

**Issue by amerkoleci** [urho3d/Urho3D#1397]
Would be nice to separate graphics backend and allow the possibility to choose on Windows platform DirectX, OpenGL or even Vulkan in the future, now we can compile only one backend per library.

**Comment by damu** [comment]
Could possible be done by linking dynamically against Urho+Backend and dynamically loading the corresponding DLL (like UrhoDirectX11.dll or UrhoOpenGL.dll) depending on the settings on application startup before loading Urho.
Would that work currently somehow? Could be easier as changing Urho to include multiple backends.

**Comment by cadaver** [comment]
Currently the graphics class headers (e.g. Texture) are subtly different in their memory layout so this would cause trouble in practice. To work properly (and be certain that things continues to work) it'd need refactoring to a base class which the API-specific implementations inherit.

There would be a runtime-penalty to pay, since it would practically force all implementation functions to be virtual.

**Comment by iSLC** [comment]
I would prefer if it stays as it is. This actually one of the things I like about Urho. It doesn't bother with fancy stuff but rather stick to being a fast and functional. There are plenty of engines out there bloated with this kind of implementation and all of them suck. @damu offered a perfectly viable solution to this.

It's not as if you're going to stop in the middle of the game and be like "_oh snap. gotta switch to dx9. oh wait, I liked opengl better. or do I? let me try one more time._" and so on...

You only chose the render device at startup which is perfectly fine to do with a custom launcher.

**Comment by dragonCASTjosh** [comment]
@iSLC I don't the use for such a feature is on the game side but more the developer side. For example I work on the graphics side alot so when I make a change I have to compile the engine 3 times and maintain duplicates of the engine. Just a pain when trying to be productive.

**Comment by cadaver** [comment]
If I'm developing / testing shaders, I typically just make copies of the Urho3DPlayer.exe (eg. UrhoD3D9, UrhoGL). Doesn't help if you actually need to make engine changes though.

**Comment by dragonCASTjosh** [comment]
@cadaver thats part of the reason i feel compiling into DLL files that can be swapped either in the application start up code or as cmd options would be extremely useful

**Comment by cadaver** [comment]
Yes, but it's a bit dubious to make a change that has negative performance impact (granted, not large) only for the sake of engine developers. Typically a project using Urho would stick to one renderer API per platform they're developing on.

**Comment by dragonCASTjosh** [comment]
i understand your point but i personally feel performance vs productivity is always a hard choice but in this case the performance inpact is so mall and the productivity games for not only engine developers but people that want to develop across platforms will be able to benefit. it also mean that if we have a renderer template then adding new backends such as DX12 and Vulkan becomes alot easier

**Comment by 1vanK** [comment]
Personally, I absolutely do not need the support of both the APIs at the same time, for the reason that I think counterproductive rewriting shaders for them when OpenGL is the most versatile. But if I would be need both of API, I will build two versions and make simple launcher (like Unity) for choice therebetween.

**Comment by dragonCASTjosh** [comment]
i may look into it once i have finished with my current PBR improvements

**Comment by damu** [comment]
Hmm should have reacted to cadavers response sooner and/or watched my mails... :-/
Typical endless loop of "ah, I'll look at that later".

I just looked at Texture.h in all three implementations. They don't look that time critical so that some indirect function calls would have a significant impact. Are there other places where it would be relevant?
There are already many virtual functions like with all those Drawables. (BTW those virtual calls could be (theoretically) completely avoided by being a bit "data oriented" and using a list for every implementations like StaticModel with all instances so that the function calls are less virtual or not virtual at all.)

Also maybe some (more) functions could be in the common base to have less virtual functions.
Also the identical data in the Texture (and others) could be reordered or moved to the base so that at least the "common part" has the same memory layout.

> It's also a potential architectural and maintenance cost, since on some platforms (eg. iOS) you likely don't want to use a renderer plugin, so there would be both static & dynamic renderer linking codepaths.
Can't Urho already be build to be linked dynamically or statically?
Or do you mean only separating the renderer into a dynamic library? I meant having one library with Urho combined with a renderer in it as that should require less work. Only the users code would have to work with Urho+different renderers without having to be rebuild. Sure that would mean having the Urho code in binary form multiple times in different library files but who cares about that bit of disk space when normally >95% are assets anyway.

**Comment by cadaver** [comment]
For a renderer plugin to be properly implemented in the traditional way also all of Graphics (SetVertexBuffer, Draw...) would need to be virtual, and these are called thousands of times per frame. Just for the Texture class the impact is minimal.

To rather make the Urho library interchangeable no matter what the graphics API could in fact be very desirable, as it could also lead to less code duplication inside the Urho codebase. In this case we'd just need to make sure the Graphics folder headers are all completely graphics API agnostic. Where necessary there could be a private implementation that changes depending on the API. Not 100% sure if this can be pulled off without trouble, but I would say this is the way to go, instead of trying to get a runtime switchable plugin and virtuals everywhere.

**Comment by damu** [comment]
Yes. The headers "normally" used by applications would have to be graphics API agnostic. There may be cases though where an application wants to directly do very special DirectX9/DirectX11/OpenGL stuff and use "private" API specific headers. These headers should of course have to be used in a way so that the user is aware that he leaves "API-independent-land" (like subfolders with the API names or class-names or namespaces).
So it's "just" about (better) separating shared code and API specific code. This should theoretically also ease adding APIs like Vulcan, Mantle or DX12 as dragonCASTjosh already mentioned.

To avoid virtual functions, function pointers, or ifdef's there could also be templates be used to have a compile-time polymorphism. Practically like ifdef's but with type safety and slightly better errors. Compile time costs should not (really) be different.

There's a BGFX library which is an graphics API agnostic abstraction for many different graphics APIs: https://github.com/bkaradzic/bgfx
I don't know how they do it but they may have a good solution. Theoretically Urho could also use BGFX (or parts of it) to outsource graphics API handling, if BGFX is usable / good enough. I've not yet used BGFX.
Does someone know BGFX? Could that be a possible better approach as doing own API abstraction?
I have the feeling that Urho and BGFX are partly quite overlapping as Urho already has some kind of API abstraction.

**Comment by dragonCASTjosh** [comment]
BGFX would be nice to have although not sure what @cadaver will think as it would require rewiring the entire renderer as BGFX handles things differently to Urho3D. I was looking into it for the Atomic guys and found it would be alot of work although after talking a little with the developer of bgfx we found it would be easiest to start from the OpenGL renderer.

BGFX is an amazing systems and i can help with building the renderer (maybe add PBR as standard, and other eyecandy) kinda treat it as a Urho3D renderer 2.0.

**Comment by damu** [comment]
> kinda treat it as a Urho3D renderer 2.0.

Sounds a bit like the overhaul the Ogre guys are doing.
Cavader mentioned elsewhere that Urho currently does the classical game engine way of doing things, in contrast to some other engines that go a more "modern" data oriented approach which gives a bit more performance but makes usage harder.
The Ogre guys are working on their overhaul for many years. Is there something else that potentially would improves Urho as well? When changing the renderer this may be an opportunity to do as well.

Is there a piece in Urho where data oriented could be kinda easily be tested out and give a significant performance impact? I kinda think there is a way to have simple usage like currently (maybe no change on the user change at all) and still be data oriented (enough) to get a speed boost.

**Comment by cadaver** [comment]
I have been away for some years from most "cutting-edge" graphics programming developments but I understand Ogre V2.x is based on the AZDO principles. In their implementation textures are put into arrays automatically by the material system, and large vertex buffers are managed by the engine to concatenate multiple meshes into. This certainly sacrifices usability for the cases where you would just want a straightforwardly functioning material/texture/vertex buffer setup, but I can't speak knowledgeably of the performance benefits.

A real world renderer has to access things like lights, occlusion, zones etc. which causes the CPU-side memory bandwidth access to grow and an "ideal" data oriented system isn't suddenly that easy to design. My gut feeling is that you would need to have many systems (physics, AI, animation, rendering) designed to work together in a data-oriented manner to start seeing noticeable gains, so e.g. experimenting with just rendering may not tell you a lot.

And finally, I think realistically for a large scale rework of Urho someone else has to take over as the project lead, because I don't think I'm going to spend anymore the (large) amount of time that would be needed for that. Consider this as an opportunity.

**Comment by dragonCASTjosh** [comment]
im happy to step in and take point on a rendering rewrite, i dont feel i know Urho enough to touch other areas

**Comment by cadaver** [comment]
In essence we already have multiple core / lead devs with their specialties, ie. me rendering & overall,  @weitjong the build system, and @hdunderscore input & web.

A rendering rewrite would not necessarily be that big if the API facing the rest of the engine is kept mostly same, but what I mean with a real rework is when the whole engine would be changed to a data-oriented / multithreaded design, at which point the things you'd have been used to, like doing a single-threaded FixedUpdate script callback from the physics simulation, would cease to exist and be transformed into something else.

**Comment by dragonCASTjosh** [comment]
How is Urho at the moment at least 7 years i think its due a modern overhall. Multi-treadding has become a core part of modern engines and modern hardware. Alot of the current urho frontend is great for users and wouldnt need much in a overhall but the backend is old, and we have a chance to bring it to modern ideals. I like the idea of a overhall

**Comment by dragonCASTjosh** [comment]
i feel that complexity will likely go up i dont think it will be anywhere near as complex as UE4, it there problem was they was focusing more on the end result then letting people use the backed. With Urho one of the core selling point is easy to access and use Source and any overall made would have to keep the source usable. 

[I think Naughty dogs system of vibers may be a good way to go for the treadding. ](http://www.gdcvault.com/play/1022186/Parallelizing-the-Naughty-Dog-Engine) it will lower complexity compared to engines like Unreal

**Comment by damu** [comment]
Whoa that took an unexpected turn... 

Ah now I get the Data Oriented Engine stuff a bit better. That's sounds like more work as I thought. I thought about using less virtual function calls (in time critical parts) by having lists of specific implementations (like StaticMesh) and saving some time with that. I guess that would not save that much when it's more about doing everything multithreaded an giant data blobs instead of small blobs per object.

I also value usability quite high. How is this instancing and StaticMeshGroup compared to Data Oriented? Could something like such optional grouping be used to have texture arrays and mesh concatenation? Would that give some AZDO / Data Oriented benefits while preserving alternatively the "easy approach"?
Do all parts have to work with this concatenation at once or would it be possible to for example let some systems like the physics one still see parts of the arrays as individual vertex buffers to safe work and upgrade part by part?

About the developer and lead developer thing: The really hard thing is to keep active. Most developer drop projects after a while or when the projects reach a certain state ("good enough for what I need").
Sure with more manpower there would be much more possible. But maybe there is a way to take small steps instead of doing a year long freeze like Ogre did. (which BTW originally moved me from Ogre to Urho).
With steps it could be like:
- introduce a more data oriented "mesh group type". (all systems still use each object in there individually like now)
- let the renderer handle the group as one big mesh to safe time and work. Possible sub-steps: Combine textures. Combine vertex buffers. ...
- more steps for physics, animation and possible other parts

**Comment by cadaver** [comment]
I believe that you'd need a real-world complex application that stresses the engine and brings out the slow spots, to see what actually needs to be improved, or can be improved. For the Ogre 2.x specific things, I actually recommend you to experiment with it (or for starters, just read their forums). Actually a year or so ago I replicated the HugeObjectCount test in Ogre 2.x and it was somewhat slower than Urho. In both engines, this stresses the culling and renderqueue by having a large number of similar objects, which the renderqueue then notices: hey same mesh and material, these can be instanced.

**Comment by dragonCASTjosh** [comment]
I may take up to work of overhauling Urho, ill likely start the end of next week but everyone is free to contribute. Also Cadaver you input as progress happens is largely welcome as it is your engine :) if things are to different to your requirements i will host it as a different "flavor" of Urho

**Comment by cadaver** [comment]
Sure, you're welcome to do that. For actually merging large changes into mainstream Urho I'd say there needs to be quite widespread argument from both users and developers that it's wanted (since it's likely that e.g. scripting is out in a completely task-based system) but you're always free to host your own fork.

**Comment by dragonCASTjosh** [comment]
i was going to post on the forums to see what people want but i seams to be down for me

**Comment by cadaver** [comment]
I suggest to be quite specific and have a plan of what exactly are you going to change, because if you post a vague plan then I guess most users would respond with a vague "yes" without understanding what is exactly going to change. Compare to when Dark Sylinc started the Ogre2.0 work, he had prepared a pdf of about 100 pages - no need to go to those lengths, but to have some idea of the scope.

Alternatively you can just start working and let the code speak for itself, if the changes are good, I'd suspect you would get users behind it even if it's not in the mainstream repo for some time being.

**Comment by dragonCASTjosh** [comment]
i way write up a small document on my plans and maybe have an area for suggestions and things we haven't considered

**Comment by cadaver** [comment]
Since this issue has veered off from the original topic quite largely, I'll close this one and make a new one for the kind of render API-refactoring we can do without worsening performance or introducing added complexity.

**Comment by bkaradzic** [comment]
This is exactly problem bgfx is solving. You can switch renderer in runtime, and engine code is not any different.

In many cases porting from GL to bgfx is straight forward. For example:

`VertexBuffer::OnDeviceReset()` <- goes away

`VertexBuffer::Release()` <- `glDeleteBuffers` becomes `bgfx::destroy<type of buffer>VertexBuffer`

`bool VertexBuffer::SetData(const void* data)` <- `glBufferData` becomes `bgfx::updateDynamicVertexBuffer`

Type of buffers are set as life span of buffer. Static buffer for meshes, dynamic buffers for things that live multiple frames (for example decals), transient buffers for short lived per frame updated things (for example particle system). All these buffers are optimizable inside bgfx without affecting API from outside.

There is also engine code simplification that happens when switching to bgfx. That's due to bgfx state tracking, draw sorting, and other things that it does for you. You don't need to worry traverse scene multiple times to get render order correctly in engine code, just submit everything at once for all passes, and then bgfx internally orders those correctly.

**Comment by dragonCASTjosh** [comment]
@bkaradzic if i was to work on porting urho to use bgfx would you assist in the effort with you being the creator of bgfx? i think the help would be incredible

**Comment by bkaradzic** [comment]
@dragonCASTjosh Sure, you can ask anything about porting...

**Comment by dragonCASTjosh** [comment]
@bkaradzic ill look into doing it once i finish up my PBR stuff

**Comment by JoshEngebretson** [comment]
bgfx is also on my "hotlist" to check into for a rendering backend.  I think this effort would be great to focus in Urho3D, if it pans out and @cadaver is into it :)  BTW, I did the bgfx hook up on the Loom engine (https://www.loomsdk.com/) back in the day...

**Comment by cadaver** [comment]
I'll be honest and put it this way that if a bgfx renderer is contributed, I'll be OK for merging it, if it appears clearly superior functionality-wise. At that point I will very likely have no interest in further Urho development, since at that point it will be little more than a scene model and a collection of third party integrations.

However, that's perfectly OK to me; you shouldn't think of Urho as "my" project any more, but rather as one belonging to the core developers ("Urho organization" in Github) and other contributors. At that point the remaining core developers should decide how to organize further development.

**Comment by 1vanK** [comment]
WIthout cadaver engine just will die :(

**Comment by JoshEngebretson** [comment]
@cadaver I am not sure what I would be doing today if it wasn't for Urho's awesome object, event, scene model, and the quality modular integrations into said model :)  Urho is a fantastic and clean library, this is exactly what I, and many others love about it!  I can truly say that Urho altered my destiny... Thank you for that, I think, we'll see where all this leads first to be sure ;)  

On bgfx, this would be a really interesting reference renderer as it has coverage across multiple backends itself.  There certainly wouldn't be any lack of interesting graphics stuff to work on, with unified shaders and without the maintenance and code replication there is now.  bgfx is still pretty low level so would take plenty of tuning and the use case would almost certainly make bgfx better too :)

**Comment by OldSnake22** [comment]
Im also afraid to see Cadaver "leaving" the project. He is doing an awesome job just by fixing issues very fast.

**Comment by weitjong** [comment]
I hope that will not happen too. I think most of us are quite happy with the current "leadership" and development direction.

**Comment by cadaver** [comment]
@weitjong I would agree that the development direction is good. I'd only step aside in the case of major architecture or rendering upheaval, if it's found to be better than the existing code and merged in. I don't think this should hinder anyone, if it actually resulted in a much better codebase for Urho3D, and further revitalized development.

**Comment by JoshEngebretson** [comment]
I am a fan of the architecture and Urho3D most definitely does not need a "rewrite".  I also believe the development direction is solid with a great deal of experience and code taste behind it.  

I like the idea of a bgfx "reference backend" starting with the OpenGL one.  This in itself would be a lot of work, though darned cool and could be good for Urho.  It wouldn't need to replace the D3D9/11/OpenGL backends.  I am not interested in "bling" from bgfx, more like the unified shader model and graphics API abstractions.  Maybe these can be achieved independently...

**Comment by cadaver** [comment]
I'm afraid that in practice bgfx integration would need to touch a lot of the high-level rendering; directly implementing the same low level Graphics API would lose a lot of the benefits of the bgfx usage model, being inefficient or maybe even impossible.

**Comment by JoshEngebretson** [comment]
Yeah, well, I got Quake2 running great on Urho, not the same thing I know ;)

This is some good source to scroll through as it gives the API a reasonable workout to see if it stirs any thoughts, https://github.com/bkaradzic/bgfx/blob/master/examples/16-shadowmaps/shadowmaps.cpp

Otherwise, unified shader code using cross compilation and sharing some of the window creation/management and whatever else can be split into common code would be good stuff on the rendering side.

**Comment by cadaver** [comment]
It's a pretty big difference. Urho's Graphics API is an imperative state machine similar to OpenGL or D3D, and when you render shadows, or do post effects, you set the appropriate rendertargets and states, then draw. Here all the different views you're going to render to are defined as data and submitted to the render queue.

The relationship of bgfx and SDL (in case we wouldn't throw away SDL) in window creation is also unclear to me, how would that go in practice?

**Comment by bkaradzic** [comment]
Seriously this should not be "low level rendering or die" style discussion. bgfx offers abstraction on low level side. It has mid-level declarative API, which separates high level and low level in the way that allows high degree of freedom to arrange low level internals in the way it works the best for given backend API, and arrange your high level scene stuff in the way it doesn't make any assumptions about low level API. Benefit of this becomes pretty obvious with D3D12 and Vulkan where typical thin abstraction wrappers designed with D3D9/11/GL are not suitable for D3D12/VK.

Anyhow losing low level scene sorting, batching / state management etc. is actually good for any engine. Your scene code becomes decoupled from graphics API, and it simplifies it. You still have to do many of high level things. This actually allow you to do more serious work on high-end graphics techniques, or state-of-art scene visibility culling, etc.

**Comment by cadaver** [comment]
I don't contest the merits of bgfx's approach. The only points I'm making are:
- It doesn't make sense to have an equal "bgfx" low level implementation in Urho alongside OpenGL / D3D9 / D3D11 while the high level rendering code stays unchanged
- I reserve the right to lose interest in development of Urho once it uses a 3rd party library for rendering (in case this is contributed and is good)

**Comment by bkaradzic** [comment]
> It doesn't make sense to have an equal "bgfx" low level implementation in Urho alongside OpenGL / D3D9 / D3D11 while the high level rendering code stays unchanged

It doesn't. It should be definately first step, to get to feature parity without breaking existing functionality. Once it's at feature parity and you can remove legacy renderers, high level stuff should be changed to work with new approach.

> I reserve the right to lose interest in development of Urho once it uses a 3rd party library for rendering (in case this is contributed and is good)

Sure, just it seems like extreme stance. 3rdparty libs used in Urho3D show high degree of pragmatism, but there is red line for renderer? :) I just think Urho3D could benefit from switching to bgfx. But from my point of view I'm totally fine that no one chooses to use bgfx. bgfx is done because there wasn't any pure open source renderer that wasn't also opinionated about underlying API. All available renderers were OpenGL only, or too high level framework forcing bunch of things onto you, or they were full engines.

**Comment by cadaver** [comment]
There are two things that must be separated: what is best for Urho project, and what personally motivates me. You possibly aren't aware, but I have almost no personal stake in Urho (e.g. no personal game projects), the motivation is fun, learning and tinkering, and the whole project started as my rendering experiments. I will not force myself to do something I do not enjoy, on the other hand the project should go ahead according to what is best for it.

**Comment by bkaradzic** [comment]
> the motivation is fun, learning and tinkering, and the whole project started as my rendering experiments

I understand that. There is nothing in bgfx that prevents you of doing it. You can think of bgfx as D3D/GL equivalent. Whatever you do there you can do in bgfx, unless you want to thinker with graphics APIs directly. But your current setup is not good for D3D12 and Vulkan, and you'll have to greatly rework how it works, probably more work than just switching to bgfx.

Horde3D had similar issue, being left behind because of renderer didn't support platforms developers wanted, but guys at DogByte create fork, ported it to bgfx and still using Horde3D to make their games: https://www.youtube.com/watch?v=CgMr1g12yXw

**Comment by nem0** [comment]
@bkaradzic  Is this Horde3D fork with bgfx available somewhere ? I would love to see that.

**Comment by bkaradzic** [comment]
@nem0 I don't think so. Search Horde3D forums to see why/how they did it: http://www.horde3d.org/forums/viewtopic.php?f=8&t=1896

**Comment by cadaver** [comment]
It's quite easy to work out the criteria: nothing obvious should break in rendering (forward / deferred paths) and there should be evidence or promise of better performance and features. And also no obvious heavy dissent from the community. I'd be happy to oversee the transition and merging.

However I cannot promise that I'd stick around in the lead position after that. Because at that point we'd essentially have added another black box to the system, similar to SDL. I know for a fact that I'd take little enjoyment from troubleshooting that. What you'd gain in development potential of new features, you could lose in visibility and understanding of the whole code path that makes the rendering happen.

**Comment by JoshEngebretson** [comment]
I would most certainly opt for @cadaver over bgfx rendereing, so there's the first voice of dissent :)

A big win would be unified shaders, this could happen without bgfx, any thoughts on this particular feature?

**Comment by 1vanK** [comment]
By the way, Unity also use third party libraries for converting shaders http://docs.unity3d.com/Manual/SL-ShadingLanguage.html

**Comment by cadaver** [comment]
@JoshEngebretson We're talking of an injection of a large piece of code after which the contributor(s) should be qualified to take over, at least the rendering part. It could be a very positive change too, but we can't know beforehand. I think the worst hypothetical case would be for them to lose interest shortly after, which is always a possibility in open source.

Yes, there are various libraries available (I'm not sure of the exact maturity) that will perform shader language conversion, and integrating a tool based on them should be doable with some effort. As long as we're talking of simple mechanical conversions, like mul() to multiplication, float2/3/4 to vec2/3/4 there should be no performance difference to handwritten.

**Comment by iSLC** [comment]
Every single game engine that I've known, which reached a certain maturity. At some point, someone came in with the idea to replace the rendering with bgfx. Any of the popular ones Godot, Ogre, Irrlicht etc. Someone sure came forward and said "_Let's demolish what you have so far and use bgfx_" Doesn't matter if it made sense or not or whether was of critical need for the person the person who recommend it. Everyone would do it for the curiosity. And probably abandon it later.

If adding bgfx would mean that @cadaver cannot be the lead developer anymore and the engine would be maintained by anyone who can. Then I'm not sure it's the best idea.

Urho is by far the best engine I've seen in a while. And I've been keeping a close eye on it for quite some time.

I've tried Irrlicht which is a pain in the @$$ to do more than hello world apps (_so to speak_). And the moment you attempt to go a little further you'll see a significant degradation in performance without doing heavy modifications to the engine itself.

Ogre is intimidating to even compile. I've spent a few good weeks just to understand how to get started with it. It's just not pretty for someone new.

Godot is nice but you're locked to their scripting language and IDE. Feels more suited for designers than programmers. To do something in C++ you have to make it a plugin/extension which means you need to compile the engine first. Which is even more intimidating than Ogre.

Which lead me to Urho. You can compile it and have something working in no time. The code is easy to understand and read through it if something is not clear or documented. The overall design is a pleasure to work with and the engine feels like it was meant to be extended.

So I'm not sure why everyone is dead set on bringing bgfx into the project when the engine just reached a certain stability and popularity. People that make AAA games will not use Urho. They'll go ahead and use a major game engine where a lot of $$$ was invested and will continue to be. (_Unity, UE4, CryEngine etc._) This engine is more for indie projects and in it's current state the engine is perfect for that.

At least release one more version before changing the graphics component. So people can fork it and use that if bgfx proves to not be what everyone is looking for :D

**Comment by 1vanK** [comment]
I tried bgfx demos today. At first sight they look interesting. But I have serious problems with the shadows for point light. They work VERY slowly (I know that it required several renderings in different directions, but in Urho no so significant performance drop). When reducing the size of the shadow map shadows look much worse than in Urho (i compared simple and pcf methods, turn on and turn off blurring). Although maybe I'm just doing something wrong.

**Comment by cadaver** [comment]
I believe you shouldn't judge bgfx based on that, as it's just demonstrating a certain (maybe not good) way to render shadows, which comes totally from the sample, bgfx itself is just about GPU resource management, state setup and draw call submission.

**Comment by cadaver** [comment]
However, when I look at the shadow sample I see a potential real engine-level problem, all shader permutations have been pregenerated (with ubershader ifdefs stripped away) and stored as files. From my limited knowledge of bgfx I don't know if this is mandatory, but something like Urho LitSolid shader has potentially hundreds of permutations which is a problem if you needed to pregenerate them all instead of just "when actually used" basis.

**Comment by nem0** [comment]
There is shader compiler that can be invoked as needed. It's not really required to pregenerate all combinations in advance. The question is whether it is the preferred way or not.

**Comment by cadaver** [comment]
Thanks. Indeed, the tougher question is platforms where you can't really invoke shaderc. This is the relevant bgfx issue, which suggests to have it as a library utility function: https://github.com/bkaradzic/bgfx/issues/156

**Comment by nem0** [comment]
Indeed, shaderc is not a library right now, but it should be no problem to create one.

**Comment by amerkoleci** [comment]
Usually is better to avoid runtime compilation as it can be slow, thus precompiled shader permutation need to be done before, this is same as Unity, Unreal, and Vulkan SPIRV is born for.

**Comment by cadaver** [comment]
There are platforms on which that isn't available (desktop GL / GLES2) so on those the best you can do is to hope your preprocessed shader source is fast for the compiler to digest.

**Comment by amerkoleci** [comment]
GL, GLES 3.0 and some GLES 2.0 supports binary shader loading, nice idea would be to have SPIRV bytecode and convert it on backend where necessary, similar to [https://github.com/KTXSoftware/krafix](url)

**Comment by JoshEngebretson** [comment]
This is a GLSL to HLSL transpiler from Microsoft: https://github.com/MicrosoftEdge/WebGL  

They recently released it, something to watch.  

Here is a HLSL to GLSL from Aras at Unity: https://github.com/aras-p/hlsl2glslfork and a GLSL optimizer used for optimizing mobile GLSL: https://github.com/aras-p/glsl-optimizer both of these are active

**Comment by JoshEngebretson** [comment]
@cadaver The work in the agnostic headers branch is looking really great and will solve a bunch of scripting issues, hope you have some good heavy metal playing :)

**Comment by cadaver** [comment]
Removal of duplicated code is also a bonus in that work, there's actually surprisingly much of it :)

**Comment by boberfly** [comment]
@bkaradzic heh, yeah I remember working with the Dogbyte guys on GL ES 2.0 support for Horde3D under the name of MistaED. A bit of a blast from the past there, back when I was tinkering with my old Nokia N900 and getting the famous golden knight to throw particles around (and realising how damn slow a simple forward-renderer was on it vs a basic single-pass with no lights). :)

@cadaver when I was toying with the idea of porting Vulkan to Urho3D I did notice basically an overhaul would need to be made to even take advantage, like with having a state-based vs passing data around and deferring stuff in a batch, a possible abstraction could be passing around structs with the right data to pass to these functions or something. I kind of veered off and started to study my own way of rendering, that enjoyment you kind of lose with going with bgfx... :(

**Comment by yushli** [comment]
I just notice this discussion when I try to search Vulkan. It may be late but still I would like to express my feelings that Urho3D definitely needs cadaver in order to move on. I am quite satisfied with the current structure and the clean code base and the technology tastes and the coding styles. If cadaver leaves Urho3D, many of us may lose faith in it. So I hope cadaver will keep interested in moving Urho3D forward and keep up the genius work.

**Comment by cadaver** [comment]
Well, I'd say Urho depends on what *everyone* brings to the table. I don't work on it as much as I used to, but I do find small things to improve here and there, and having it run correctly (within limits of reasonable effort -- there are always odd platforms and corner cases) is important to me.

However I don't recommend expecting major changes from me at this point, since you're likely to end up disappointed. Like I stated before, in the hypothetical scenario that someone turns up with strong detailed plans of what Urho would need to do to sort of "level up", and with proof that they have the skill and energy for it, I'd be glad to let them take over the lead role and it would likely be for the project's best. Reality shows that this is a highly hypothetical scenario, but it has happened in Ogre, when dark_sylinc took over.

**Comment by yushli** [comment]
If that hypothetical scenario do happen, they should fork Urho3D out and make a new one instead of changing this project and taking over the lead role from you. As many have already said, Urho3D cannot go on without the original genius. If someone comes up with significant changes or so called "big improvements", it should be applied by forking out and see if others do like the technology choices and decisions. Even if the changes are fine, a new project should be created instead of breaking this Urho3D. This Urho3D I only trust the original genius(cadaver)'s dedication, technology tastes and trade off decisions.

**Comment by OldSnake22** [comment]
@boberfly I see that you are working actively on BGFX implementation and this is great ! Can you provide me some information about when do you think it will be ready ? What is still left to do ? Thank and keep up your good work.

**Comment by rokups** [comment]
Hey @OldSnake22. Just today @boberfly managed to get some visible result - a red screen, so it is quite a while away. All i know is that @boberfly seems to be about finishing shader stuff at the moment. He can comment more in depth when he wakes up.

**Comment by boberfly** [comment]
Hey, yea just smashing at the debugger right now, pretty much everything is implemented in some wip form now, including shaders but just trying to get the hello world to render right now. Maybe some time this week we'll get something resembling an alpha? Then once I progress to the more complicated examples I can stress-test the more complex render paths/framebuffer stuff and fix up any shader errors. 

I'm getting GL3/DX11 to work first, then I'll need to bring over the checks for mobile/GLES2/GL2 so they work as intended and then polish++

**Comment by OldSnake22** [comment]
Perfect ! Thank you both for your answers. This will be a nice christmas present ;)

**Comment by boberfly** [comment]
![urho_bgfx](https://user-images.githubusercontent.com/5601445/33312278-3aaa3790-d3dc-11e7-8354-fe887ccfdb32.png)

**Comment by OldSnake22** [comment]
Nice im sure that you will have a hello world working example very soon ;) Can you keep us informed about the progress on this thread as you make some progress ? Thank you.

Beside this BGFX base implementation are you planning to integrate some features like geometry / compute shaders for example ?

**Comment by boberfly** [comment]
Still plugging away btw with small fixes here and there. The red is gone (.Uint() on a colour is a different order so alpha became red). I am stuck right now with a few things and renderdoc doesn't give me enough info.

**Comment by boberfly** [comment]
![urho_bgfx4](https://user-images.githubusercontent.com/5601445/33468785-a0e8bee4-d613-11e7-9da1-e6db62599dea.png)

Yay screenshot...
https://github.com/boberfly/Urho3D/releases/tag/0.0

If you want to make renderDoc log dumps, make a temp dir in Debug and hit F11...

**Comment by OldSnake22** [comment]
Really hope you can sort it out :) Good luck !

**Comment by boberfly** [comment]
![urho_bgfx5](https://user-images.githubusercontent.com/5601445/33473573-d326535e-d62b-11e7-80c5-e24dec1c3b30.png)

**Comment by boberfly** [comment]
![urho_bgfx6](https://user-images.githubusercontent.com/5601445/33707093-16360d4c-daeb-11e7-81af-b7f588fa4382.png)
bgfx dx11 vs urho dx11

**Comment by eugeneko** [comment]
@boberfly Great! I must test it. But... where is FPS?!

**Comment by OldSnake22** [comment]
Congrats ! +1 for the FPS comparison please :)

**Comment by bkaradzic** [comment]
@eugeneko FPS is on screenshot on the left (actually a lot of info there, for example red text means it's debug build)

**Comment by aurodev** [comment]
2320.13 FPS for a single instanced cube  🥇    Lol!  

Just getting this into/underneath Urho is awesome!  Good work!
Hopefully the actual changes to the API, xplat etc., when this is integrated, will be minimal. 

@boberfly  @bkaradzic  Thanks!  

#urhorocks

**Comment by bkaradzic** [comment]
@aurodev Yeah this scene is not really good perf test, + he has 1080 which is quite high end.

**Comment by boberfly** [comment]
Yep no benchmarks yet, instancing is not in yet plus I need to fix things first like the projection matrix (the moment you move the camera it all breaks)... :)

**Comment by 1vanK** [comment]
Is it works as additional layer?

Standart: Engine > Graphics API > OpengGL, DirectX
With Bgfx: Engine > Graphics API > Bgfx > OpengGL, DirectX, Vulkan

**Comment by bkaradzic** [comment]
Nope, what's there "Graphics API" becomes bgfx, and what's scene becomes simplified because switch from imperative to declarative API, and functionality that's moved inside bgfx (sorting, state tracking, etc).

**Comment by 1vanK** [comment]
I'm not sure I completely understand. Scene node system will be changed?

**Comment by bkaradzic** [comment]
Not user facing / high level, but internals / guts, since it does redundant stuff. Output result should not change.

**Comment by 1vanK** [comment]
> since it does redundant stuff

for example?

**Comment by bkaradzic** [comment]
See above:

> and functionality that's moved inside bgfx (sorting, state tracking, etc).

**Comment by 1vanK** [comment]
Sorting of batches is done in several lines. Why use for this additional library?

p.s. I somehow thought that in bgfh have Vulkan, but now I looked into the repositories and did not find it

**Comment by bkaradzic** [comment]
You get all Urho rendering features already supports + D3D12, Metal, (and Vulkan after it's done: https://github.com/bkaradzic/bgfx/blob/master/src/renderer_vk.cpp)

Then you get ability to switch rendering backend, choose GPU (for multi-GPU machines) in runtime (see issue title above).

You get unified shader language, so authoring shaders is done once, not once per rendering backend.

You get ability to submit rendering draw calls from multiple threads.

Switch from imperative to declarative API allows you to simplify scene graph, so you don't need to iterate multiple times over the same objects, and can submit all passes at once during single pass.

You actually get someone to maintain renderers, since project lost someone who was doing that.

**Comment by 1vanK** [comment]
Thanks for clarifying

**Comment by boberfly** [comment]
Currently it is GraphicsAPI > BGFX > DX/GL/etc. so not much changes at all to keep the other ones working side by side, however there are parts which are bypassed like the sorting found in Batch.cpp is now delegated to BGFX, and any GL/DX-specific parts are now checked using caps.

**Comment by dragonCASTjosh** [comment]
imagine it is set up this way as an initial pass so its easier to get things up and running. In the long run do you plan to remove Urhos abstraction layer in favor of direct BGFX access?
